design/jwt-verification-design.md

```diff
@@ -0,0 +1,205 @@
+
+A complete/valid HTTPProxy using JWT verification is shown below:
+
+```yaml
+apiVersion: projectcontour.io/v1
+kind: HTTPProxy
+metadata:
+  name: jwt-verification-proxy
+spec:
+  virtualhost:
+    fqdn: example.com
+    tls:
+      secretName: tls-cert
+    jwtProviders:
+      - name: provider-1
+        issuer: example.com
+        audiences:
+          - audience-1
+          - audience-2
+        remoteJWKS:
+          httpURI:
+            uri: https://example.com/jwks.json
+            timeout: 1s
+          cacheDuration: 5m
+  routes:
+    - conditions:
+        - prefix: /
+      jwtProvider: provider-1
+      services:
+        - name: s1
+          port: 80
+    - conditions:
+        - prefix: /css
+      services:
+        - name: s1
+          port: 80
+```
+
+
+## Alternatives Considered
+
+### OAuth2 filter
+Envoy also has an [OAuth2 HTTP filter](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/oauth2_filter), which supports end-to-end OAuth2/OIDC flows.
+This provides related but separate functionality to the JWT authentication filter.
+This [excellent blog post](https://www.jpmorgan.com/technology/technology-blog/protecting-web-applications-via-envoy-oauth2-filter) shows an example of how to use the OAuth2 and JWT filters together in Envoy.
+Contour [may pursue adding support for the OAuth2 filter](https://github.com/projectcontour/contour/issues/2664), but it will be designed and implemented separately.
+
+
+## Defining JWT verification rules separately from routes
+
+An alternate API design is to define JWT verification rules separately from routing rules, as shown in the below YAML:
+
+```yaml
+apiVersion: projectcontour.io/v1
+kind: HTTPProxy
+metadata:
+  name: jwt-verification-proxy
+spec:
+  virtualhost:
+    fqdn: example.com
+    tls:
+      secretName: tls-cert
+    jwtVerificationPolicy:
+      providers:
+        - name: provider-1
+          issuer: example.com
+          audiences:
+            - audience-1
+            - audience-2
+          remoteJWKS:
+            httpURI:
+              uri: https://example.com/jwks.json
+              timeout: 1s
+            cacheDuration: 5m
+      rules:
+        - match:
+            prefix: /js
+        - match:
+            prefix: /
+          providerName: provider-1
+  routes:
+    - conditions:
+        - prefix: /
+      services:
+        - name: s1
+          port: 80
+```
+
+This design has the potential benefit of reducing duplication in the routing rules, by not requiring an entire route to be duplicated in order to express an "except" condition for JWT verification.
+It also allows the user to have direct control over the order in which JWT rules are applied, versus being subject to Contour's rules for sorting routes.
+
+However, this design results in two separate places where route-related behavior are defined, and also creates cognitive overhead for users by having two separate ways of ordering match criteria.
```

Comment:
Another question: I have not studied `jwt_authn` filter at this level before. Are these two API alternatives also reflecting the alternatives where we can add `jwt_authn` filter in the Envoy config? 

I did not see full example in Envoy docs but I guess it could be added in various places? It could be added to `http_filters` of `HttpConnectionManager` , or to `typed_per_filter_config` of `config.route.v3.VirtualHost `, or `typed_per_filter_config` of `config.route.v3.Route`

Requirements uses `RouteMatch` to apply verification, similarly as `Route` applies them for routing requests. Does this imply following: if `jwt_authn` filter is added inside `config.route.v3.Route`, which itself already has matched the URL for routing, the match requirement inside  `jwt_authn` does not matter much anymore?  The requirement could simply match `/` since route level matching has already happened before we reach `jwt_authn` filter . On the other hand, if `jwt_authn` is used inside virtualhost (before routing), it requires duplicating all match rules that need to be also present in virtualhost's Routes?

Reply:
> On the other hand, if jwt_authn is used inside virtualhost (before routing), it requires duplicating all match rules that need to be also present in virtualhost's Routes?

yep, exactly, which is why IMO reducing duplication and keeping the opt-in/out of JWT verification for a Route is a bit cleaner, we can reduce what users need to configure and simply have them use the existing route matching rules

Reply:
Yeah, this is a good point, having the JWT providers specified on the routing rules themselves should actually simplify how we configure Envoy since we can put the configuration in the routes themselves in per-route config. I need to play around with it some since it's not how I initially implemented it, and the docs are not super-detailed, but definitely a good though to pursue.

Reply:
yeah the per-route config seems like where you can set what requirement you want to be verified (by name), but the definition of the requirements/providers still has to go in the filter

Reply:
I made this switch in my WIP branch and it seems to be working well. The other nice benefit it has is it avoids us needing to separately sort a list of JWT requirements -- now the sorting happens as part of the normal route sorting we do. Helps avoid some ugly code duplication in our sorter package.